Fails to link wasm32 binaries.

```
CMake Error at /nix/store/sgsvy949kd2sr2vsxcffz5jkv855rmib-cmake-3.22.2/share/cmake-3.22/Modules/CMakeTestCCompiler.cmake:69 (message):
  The C compiler

    "/nix/store/lmnvxwbzh8nbxls0i8604bjk6pv5fygg-wasm32-unknown-wasi-clang-wrapper-8.0.1/bin/wasm32-unknown-wasi-clang"

  is not able to compile a simple test program.

  It fails with the following output:

    Change Dir: /build/libcxxabi-8.0.1.src/build/CMakeFiles/CMakeTmp
    
    Run Build Command(s):/nix/store/is8wg3j77z9ndpsvkm85sfbrlvvqd66l-gnumake-4.3/bin/make -f Makefile cmTC_387d9/fast && /nix/store/is8wg3j77z9ndpsvkm85sfbrlvvqd66l-gnumake-4.3/bin/make  -f CMakeFiles/cmTC_387d9.dir/build.make CMakeFiles/cmTC_387d9.dir/build
    make[1]: Entering directory '/build/libcxxabi-8.0.1.src/build/CMakeFiles/CMakeTmp'
    Building C object CMakeFiles/cmTC_387d9.dir/testCCompiler.c.obj
    /nix/store/lmnvxwbzh8nbxls0i8604bjk6pv5fygg-wasm32-unknown-wasi-clang-wrapper-8.0.1/bin/wasm32-unknown-wasi-clang    -MD -MT CMakeFiles/cmTC_387d9.dir/testCCompiler.c.obj -MF CMakeFiles/cmTC_387d9.dir/testCCompiler.c.obj.d -o CMakeFiles/cmTC_387d9.dir/testCCompiler.c.obj -c /build/libcxxabi-8.0.1.src/build/CMakeFiles/CMakeTmp/testCCompiler.c
    clang-8: warning: argument unused during compilation: '-rtlib=compiler-rt' [-Wunused-command-line-argument]
    Linking C executable cmTC_387d9
    /nix/store/sgsvy949kd2sr2vsxcffz5jkv855rmib-cmake-3.22.2/bin/cmake -E cmake_link_script CMakeFiles/cmTC_387d9.dir/link.txt --verbose=1
    /nix/store/lmnvxwbzh8nbxls0i8604bjk6pv5fygg-wasm32-unknown-wasi-clang-wrapper-8.0.1/bin/wasm32-unknown-wasi-clang CMakeFiles/cmTC_387d9.dir/testCCompiler.c.obj -o cmTC_387d9 
    wasm32-unknown-wasi-wasm-ld: error: /nix/store/fi3903r6zj9psp8c9ry6llq0lw0j9fv2-wasilibc-static-wasm32-unknown-wasi-unstable-2021-09-23/lib/libc.a(dlmalloc.o): Bad section type
    clang-8: error: lld command failed with exit code 1 (use -v to see invocation)
    make[1]: *** [CMakeFiles/cmTC_387d9.dir/build.make:100: cmTC_387d9] Error 1
    make[1]: Leaving directory '/build/libcxxabi-8.0.1.src/build/CMakeFiles/CMakeTmp'
    make: *** [Makefile:127: cmTC_387d9/fast] Error 2
```

--- a/pkgs/applications/networking/browsers/firefox/common.nix
+++ b/pkgs/applications/networking/browsers/firefox/common.nix
@@ -15,7 +15,9 @@
 , rust-cbindgen, nodejs, nasm, fetchpatch
 , gnum4
 , gtk3, wrapGAppsHook
+, pkgsCross
 , debugBuild ? false
+, runCommand
 
 ### optionals
 
@@ -121,6 +123,15 @@ let
                 })
                 else stdenv;
 
+  # Compile the wasm32 sysroot to build the RLBox Sandbox
+  # https://hacks.mozilla.org/2021/12/webassembly-and-back-again-fine-grained-sandboxing-in-firefox-95/
+  # We only link c++ libs here, our compiler wrapper can find wasi libc and crt itself.
+  wasiSysRoot = runCommand "wasi-sysroot" {} ''
+    mkdir -p $out/lib/wasm32-wasi
+    for lib in ${pkgsCross.wasi32.llvmPackages.libcxx}/lib/* ${pkgsCross.wasi32.llvmPackages.libcxxabi}/lib/*; do
+      ln -s $lib $out/lib/wasm32-wasi
+    done
+  '';
 in
 
 buildStdenv.mkDerivation ({
@@ -213,7 +224,11 @@ buildStdenv.mkDerivation ({
     configureScript="$(realpath ./mach) configure"
     export MOZBUILD_STATE_PATH=$(pwd)/mozbuild
 
-  '' + (lib.optionalString googleAPISupport ''
+  '' + (lib.optionalString (lib.versionAtLeast version "95.0") ''
+    # RBox WASM Sandboxing
+    export WASM_CC=${pkgsCross.wasi32.stdenv.cc}/bin/${pkgsCross.wasi32.stdenv.cc.targetPrefix}cc
+    export WASM_CXX=${pkgsCross.wasi32.stdenv.cc}/bin/${pkgsCross.wasi32.stdenv.cc.targetPrefix}c++
+  '') + (lib.optionalString googleAPISupport ''
     # Google API key used by Chromium and Firefox.
     # Note: These are for NixOS/nixpkgs use ONLY. For your own distribution,
     # please get your own set of keys.
@@ -258,6 +273,7 @@ buildStdenv.mkDerivation ({
   ++ lib.optional ltoSupport "--enable-lto=cross" # Cross-language LTO.
   ++ lib.optional (ltoSupport && (buildStdenv.isAarch32 || buildStdenv.isi686 || buildStdenv.isx86_64)) "--disable-elf-hack"
   ++ lib.optional (ltoSupport && !buildStdenv.isDarwin) "--enable-linker=lld"
+  ++ lib.optional (lib.versionAtLeast version "95") "--with-wasi-sysroot=${wasiSysRoot}"
 
   ++ flag alsaSupport "alsa"
   ++ flag pulseaudioSupport "pulseaudio"
@@ -277,7 +293,6 @@ buildStdenv.mkDerivation ({
   ++ lib.optionals enableDebugSymbols [ "--disable-strip" "--disable-install-strip" ]
 
   ++ lib.optional enableOfficialBranding "--enable-official-branding"
-  ++ lib.optional (lib.versionAtLeast version "95") "--without-wasm-sandboxed-libraries"
   ++ extraConfigureFlags;
 
   postConfigure = ''
@@ -356,6 +371,7 @@ buildStdenv.mkDerivation ({
     inherit applicationName;
     inherit tests;
     inherit gtk3;
+    inherit wasiSysRoot;
   } // extraPassthru;
 
   hardeningDisable = [ "format" ]; # -Werror=format-security

--- a/pkgs/development/libraries/wasilibc/default.nix
+++ b/pkgs/development/libraries/wasilibc/default.nix
@@ -1,20 +1,35 @@
-{ stdenv, fetchFromGitHub, lib }:
+{ stdenv, buildPackages, fetchFromGitHub, lib }:
 
 stdenv.mkDerivation {
   pname = "wasilibc";
-  version = "20190712";
-  src = fetchFromGitHub {
-    owner = "CraneStation";
+  version = "unstable-2021-09-23";
+
+  src = buildPackages.fetchFromGitHub {
+    owner = "WebAssembly";
     repo = "wasi-libc";
-    rev = "8df0d4cd6a559b58d4a34b738a5a766b567448cf";
-    sha256 = "1n4gvgzacpagar2mx8g9950q0brnhwz7jg2q44sa5mnjmlnkiqhh";
+    rev = "ad5133410f66b93a2381db5b542aad5e0964db96";
+    hash = "sha256-RiIClVXrb18jF9qCt+5iALHPCZKYcnad7JsILHBV0pA=";
+    fetchSubmodules = true;
   };
+
+  # clang-13: error: argument unused during compilation: '-rtlib=compiler-rt' [-Werror,-Wunused-command-line-argument]
+  postPatch = ''
+    substituteInPlace Makefile \
+      --replace "-Werror" ""
+  '';
+
+  preBuild = ''
+    export NIX_CFLAGS_COMPILE="-I$(pwd)/sysroot/include $NIX_CFLAGS_COMPILE"
+  '';
+
   makeFlags = [
     "WASM_CC=${stdenv.cc.targetPrefix}cc"
     "WASM_NM=${stdenv.cc.targetPrefix}nm"
     "WASM_AR=${stdenv.cc.targetPrefix}ar"
     "INSTALL_DIR=${placeholder "out"}"
   ];
+
+  enableParallelBuilding = true;
 
   postInstall = ''
     mv $out/lib/*/* $out/lib
@@ -23,9 +38,9 @@ stdenv.mkDerivation {
 
   meta = with lib; {
     description = "WASI libc implementation for WebAssembly";
-    homepage    = "https://wasi.dev";
-    platforms   = platforms.wasi;
-    maintainers = [ maintainers.matthewbauer ];
+    homepage = "https://wasi.dev";
+    platforms = platforms.wasi;
+    maintainers = with maintainers; [ matthewbauer ];
     license = with licenses; [ asl20 mit llvm-exception ];
   };
 }

--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -12906,7 +12906,7 @@ with pkgs;
       else if platform.isFreeBSD then 7
       else if platform.isAndroid then 12
       else if platform.isLinux then 11
-      else if platform.isWasm then 8
+      else if platform.isWasm then 12
       else latest_version;
     # We take the "max of the mins". Why? Since those are lower bounds of the
     # supported version set, this is like intersecting those sets and then

